### ISSUE TITLE
`reportUnnecessaryIsInstance=error` is missing all "X is never an instance of Y" cases

### ISSUE THREAD

**Issue by bukzor** [microsoft/pyright#3789]
**Describe the bug**

All three of these conditionals are statically unnecessary, but only the second is reported:

```python
# pyright: off, reportUnnecessaryIsInstance=true                                                                                                              

x: int = 3                                                                                                                                                    
                                                                                                                                                              
if x is None:                                                                                                                                                 
    print("is None")                                                                                                                                          
elif isinstance(x, int):  # E: Pyright: Unnecessary isinstance call; "Literal[3]" is always an instance of "int"                                               
    print("is int")                                                                                                                                           
elif isinstance(x, str):                                                                                                                                      
    print("is str")                                                                                                                                           
```

**To Reproduce**

Run pyright on the above code.

**Expected behavior**

I expect _all_ three conditionals to be flagged as unnecessary.

**Comment by erictraut** [comment]
This is working as intended. The first check does not involve an `isInstance` conditional, so the `reportUnnecessaryIsInstance` doesn't apply.

In the third case, the error is suppressed because the type of `x` has already been narrowed to `Never` previously. If you fix or remove the second unnecessary conditional, the third will then be flagged.

**Comment by erictraut** [comment]
Actually, the third check is not unnecessary. It's possible for a type to be a subclass of both `int` and `str`, so `isinstance(x, str)` may not be unnecessary.

**Comment by bukzor** [comment]
Please to reopen. Here's a demonstration you'll find more agreeable:

```
# pyright: off, reportUnnecessaryIsInstance=true                                                                                                              

x: int = 3                                                                                                                                                    
                                                                                                                                                              
if isinstance(x, str):                                                                                                                                      
    print("is str")
```

**Comment by erictraut** [comment]
Here's an example of why `reportUnnecessaryIsInstance` should not flag this as an error.

```python
# pyright: off, reportUnnecessaryIsInstance=true

class IntOrStr(int, str):
    ...

def func(x: int):
    if isinstance(x, str):
        print("is str")

func(IntOrStr())
```

**Comment by bukzor** [comment]
Since None is the only member of NoneType, the idiom in python for `isinstance(x, NoneType)` is written as `x is None`, so it's appropriate to treat `is None` as a type assertion. Isn't that exactly how `is None` is treated during type narrowing.

But I'll move this discussion to a separate issue if you believe it's off topic.

**Comment by bukzor** [comment]
So is it true to say that pyright is always unable to infer `variable X never has type Y`? There's no way to denote this?

That's my last question. Thank you for your time!

**Comment by bukzor** [comment]
It's actually impossible to subclass int and str:

```
>>> class IntOrStr(int, str):
...   pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: multiple bases have instance lay-out conflict
```

For user-defined types X and Y, I take your point, but in this case it's unnecessary to check `isinstance(str)`.

If you don't care to add such esoteric knowledge to pyright, that's another matter :)

**Comment by erictraut** [comment]
Statically proving that variable `x` _does not_ derive from type `Y` is generally not possible because of inheritance. It is possible to statically determine when `x` _does_ derive from type `Y`, so `reportUnnecessaryIsInstance` is able to catch cases where the conditional check is redundant, such as the second conditional check in the example at the top of this thread.

The `reportUnnecessaryIsInstance` check is not meant to be bulletproof. It is meant to catch common programming errors, not detect every possible unnecessary isinstance call.

**Comment by bukzor** [comment]
Thank you again. Would you accept a bug noting that pyright cannot infer that (str, int, None) and friends are disjoint types?

**Comment by erictraut** [comment]
That would require very special-case knowledge about these types that would need to be hard-coded into pyright. We generally try to avoid this because it's just not scalable to maintain such hard-coded knowledge. If this were a source of really common bugs, I might consider it, but I don't think it is.

**Comment by bukzor** [comment]
Is there a less bothersome place that I can chat about these things? I get the sense that this was an unwelcome bug report.

For example, I'm probably wrong, but the type *widening* happening here seems unintentional:

```
# pyright: off, reportUnnecessaryIsInstance=true                                                                                                              
                                                                                                                                                              
x: int = 3                                                                                                                                                    
                                                                                                                                                              
reveal_type(x)  # I: Pyright: Type of "x" is "Literal[3]"      pylint: Undefined variable 'reveal_type'                                                        
if isinstance(x, str):                                                                                                                                        
    print("is str")                                                                                                                                           
if isinstance(x, int):  # E: Pyright: Unnecessary isinstance call; "<subclass of int and str> | Literal[3]" is always an instance of "int"
    print("is int")
```

**Comment by erictraut** [comment]
There is a discussions forum if you have questions about pyright and want to understand why it behaves the way it does.

**Comment by erictraut** [comment]
Also, if your questions are not specific to pyright and apply more generally to the Python type system, the [python/typing discussion forum](https://github.com/python/typing/discussions) is also a good place to post those questions.
